restapi.go

```diff
@@ -2566,3 +2576,222 @@ func (s *Session) FollowupMessageEdit(appID string, interaction *Interaction, me
 }
+
+////////////////////////////////
+//////////  THREADS  //////////
+//////////////////////////////
+
+// StartThreadWithMessage creates a new public thread from an existing message.
+// When called on a ChannelTypeGuildText channel, creates a ChannelTypeGuildPublicThread.
+// When called on a ChannelTypeGuildNews channel, creates a ChannelTypeGuildNewsThread.
+// The id of the created thread will be the same as the id of the message,
+// and as such a message can only have a single thread created from it.
+//
+// POST /channels/{channel.id}/messages/{message.id}/threads
+func (s *Session) StartThreadWithMessage(channelID, messageID string, data *ThreadCreateData) (st *Channel, err error) {
+	body, err := s.RequestWithBucketID("POST", EndpointChannelMessageThreads(channelID, messageID), data, EndpointChannelMessageThreads("", ""))
+	if err != nil {
+		return
+	}
+
+	err = unmarshal(body, &st)
+	return
+}
+
+// StartThreadWithoutMessage creates a new thread that is not connected
+// to an existing message.
+// The created thread is always a ChannelTypeGuildPrivateThread
+//
+// POST /channels/{channel.id}/threads
+func (s *Session) StartThreadWithoutMessage(channelID string, data *ThreadCreateData) (st *Channel, err error) {
+	body, err := s.RequestWithBucketID("POST", EndpointThreads(channelID), data, EndpointThreads(""))
+	if err != nil {
+		return
+	}
+
+	err = unmarshal(body, &st)
+	return
+}
+
+// JoinThread joins the session user to a thread
+// PUT/channels/{channel.id}/thread-members/@me
+func (s *Session) JoinThread(channelID string) (err error) {
+	_, err = s.RequestWithBucketID("PUT", EndpointThreadMembers(channelID)+"@me", nil, EndpointThreadMembers(""))
+	return
+}
+
+// AddUserToThread adds another user to a thread
+// PUT /channels/{channel.id}/thread-members/{user.id}
+func (s *Session) AddUserToThread(channelID, uID string) (err error) {
+	_, err = s.RequestWithBucketID("PUT", EndpointThreadMembers(channelID)+uID, nil, EndpointThreadMembers(""))
+	return
+}
+
+// LeaveThread removes the session user from a thread
+// DELETE /channels/{channel.id}/thread-members/@me
+func (s *Session) LeaveThread(channelID string) (err error) {
+	_, err = s.RequestWithBucketID("DELETE", EndpointThreadMembers(channelID)+"@me", nil, EndpointThreadMembers(""))
+	return
+}
+
+// RemoveUserFromThread removes another user from a thread
+// PUT /channels/{channel.id}/thread-members/{user.id}
+func (s *Session) RemoveUserFromThread(channelID, uID string) (err error) {
+	_, err = s.RequestWithBucketID("DELETE", EndpointThreadMembers(channelID)+uID, nil, EndpointThreadMembers(""))
+	return
+}
+
+// ThreadMembers lists the members of a thread
+// GET /channels/{channel.id}/threads-members
+func (s *Session) ThreadMembers(channelID string) (st []*ThreadMember, err error) {
+	body, err := s.RequestWithBucketID("GET", EndpointThreadMembers(channelID), nil, EndpointThreadMembers(""))
+	if err != nil {
+		return
+	}
+
+	err = unmarshal(body, &st)
+	return
+}
```

Comment:
It doesn't look like this function works at all:
```go
panic: HTTP 404 Not Found, {"message": "404: Not Found", "code": 0}
```
It seems like it's calling to `/thread-members/` rather than `/thread-members`. The fact that `EndpointThreadMembers` doesn't take in the user ID as a parameter seems like a bit of a smell. Having two separate endpoint functions and taking in the user ID as a parameter in one (and consequently updating `JoinThread`, `AddUserToThread`, etc. might be a good idea.

Reply:
Updated